### PR TITLE
Patch ruby to support MSVC 2019 and allow win32 build

### LINKED ISSUE #25: Revert ruby_installer to bincrafters/stable
Had to build my own ruby_installer when updating to use Conan center index (cf https://github.com/NREL/conan-openstudio-ruby/pull/20 based on https://github.com/jmarrec/conan-openstudio-ruby/pull/4)

My PR https://github.com/bincrafters/conan-ruby_installer/pull/7 got merged now.

Change back to bincrafters/stable:

https://github.com/NREL/conan-openstudio-ruby/blob/705e89f9f7f87f00b59b59d2e2b90409b0cceb9f/CMakeLists.txt#L67

I'm not doing it right now because it involves a lengthy process (travis and appveyor to build over the course of several hours, making sure NREL/OpenStudio doesn't break in the process) and it works fine now.

### LINKED ISSUE #26: Patch source win32/Makefile.sub to use MSC_VER range instead of hardcoded one
`win32/Makefile.sub#538`

```
#if _MSC_VER != $(MSC_VER)
#error MSC version unmatch: _MSC_VER: $(MSC_VER) is expected.
#endif
```

MSVC 2019 has the range from 1920 to 1925 (currently), our current appveyor-built binary was on 1924, so if you use the latest MSVC 2019, you hit a hard error.

| Abbreviation  | Product name [Visual Studio version] †1 | VC ++ version †2 | _MSC_VER | _MSC_FULL_VER |
|---------------|-----------------------------------------|------------------|----------|---------------|
| 2019 Update 5 | Visual Studio 2019 version 16.5.1       | 14.25            | 1925     | 192528611     |
| 2019 Update 4 | Visual Studio 2019 version 16.4.0       | 14.24            | 1924     | 192428314     |
| 2019 Update 3 | Visual Studio 2019 version 16.3.2       | 14.21            | 1923     | 192328105     |
| 2019 Update 2 | Visual Studio 2019 version 16.2.3       | 14.21            | 1922     | 192227905     |
| 2019 Update 1 | Visual Studio 2019 version 16.1.2       | 14.21            | 1921     | 192127702     |
| 2019          | Visual Studio 2019 version 16.0.0       | 14.20            | 1920     | 192027508     |
| 2017 Update 9 | Visual Studio 2017 version 15.9.11      | 14.16            | 1916     | 191627030     |

### PR DESCRIPTION
**Pull request by jmarrec**
Fix #26 - patch MSVC ranges
Fix #25 - revert ruby installers to bincrafters/stable

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -260,24 +265,29 @@ else()
...
       COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI/src/x86/ffitarget.h ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI-install/$<CONFIG>/include/ffitarget.h
       COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI-install/$<CONFIG>/lib
-      COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI/win32/vc14_x64/x64/$<CONFIG>/libffi.lib ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI-install/$<CONFIG>/lib/libffi.lib
-      #COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI/win32/vc14_x64/x64/$<CONFIG>/libffi.pdb ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI-install/$<CONFIG>/lib/libffi.pdb
+      COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI/win32/vc14_${ARCH}/${FFI_LIB_FOLDER}$<CONFIG>/libffi.lib ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI-install/$<CONFIG>/lib/libffi.lib
+      #COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI/win32/vc14_${ARCH}/${FFI_LIB_FOLDER}$<CONFIG>/libffi.pdb ${CMAKE_BINARY_DIR}/FFI-prefix/src/FFI-install/$<CONFIG>/lib/libffi.pdb
```
Comment (jmarrec) [inline CMakeLists.txt:295]: CMake-Foo to allow building 32 bit on windows.

File: appveyor.yml [hunk 2]
```diff
@@ -3,24 +3,33 @@ build: false
...
-    CONAN_ARCHS: x86_64   # Don't want to build for x86
+    # CONAN_ARCHS: x86_64,   # Don't want to build for x86
 
     matrix:
 
+        # Try a Release build on 32bit
+        - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2019
+          CONAN_VISUAL_VERSIONS: 16
+          CONAN_BUILD_TYPES: Release
+          CONAN_ARCHS: x86
```
Comment (jmarrec) [inline appveyor.yml]: Add a CI build for windows 32 bit, release only.

File: build_ruby.bat.in [hunk 3]
```diff
@@ -60,14 +60,16 @@ IF "%LIBYAML_DIR%" == "" (
...
+    set CONF_ARGS=%CONF_ARGS% --target=x64-mswin64"
+)
 echo "New Flags %CFLAGS%"
 echo "EXTLIBS='%EXTLIBS%'"
 echo "CONF_ARGS='%CONF_ARGS%'"
 
 @ECHO ON
 
-call ${CMAKE_BINARY_DIR}\Ruby-prefix\src\Ruby\win32\configure.bat --with-static-linked-ext %CONF_ARGS% --without-win32ole --disable-install-doc --prefix="${CMAKE_BINARY_DIR}/Ruby-prefix/src/Ruby-install/" --target=x64-mswin64 --with-baseruby="%BASE_RUBY%" 
+call ${CMAKE_BINARY_DIR}\Ruby-prefix\src\Ruby\win32\configure.bat --with-static-linked-ext %CONF_ARGS% --without-win32ole --disable-install-doc --prefix="${CMAKE_BINARY_DIR}/Ruby-prefix/src/Ruby-install/" --with-baseruby="%BASE_RUBY%"
```
Comment (jmarrec) [inline build_ruby.bat.in:72]: Update the build_ruby.bat script to allow building on 32 bit.

File: conanfile.py [hunk 4]
```diff
@@ -263,63 +264,72 @@ def package_info(self):
...
+                              'struct-{l}.lib',
+                              'symbol-{l}.lib',
+                              'thread_fd_close-{l}.lib',
+                              'time-{l}.lib',
+                              'tracepoint-{l}.lib',
+                              'typeddata-{l}.lib',
+                              'update-{l}.lib',
+                              'usr-{l}.lib',
+                              'wait_for_single_fd-{l}.lib']
+            expected_libs = [x.format(l=libarch) for x in expected_libs]
```
Comment (jmarrec) [inline conanfile.py:356]: Update the code to gather libs to take 32 bit into account

File: patches/0001-patch-to-support-version-ranges-for-MSVC.patch [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ #if _MSC_VER != $(MSC_VER)
+ #error MSC version unmatch: _MSC_VER: $(MSC_VER) is expected.
+ #endif
++!endif
+ #define RUBY_MSVCRT_VERSION $(RT_VER)
+ #define STDC_HEADERS 1
+ #define HAVE_SYS_TYPES_H 1
+-- 
+2.21.0.windows.1
+
```
Comment (jmarrec) [inline patches/0001-patch-to-support-version-ranges-for-MSVC.patch:36]: Patch to the ruby win32/Makefile that has macro that doesn't expect MSVC 2019 (while it works fine), avoid hitting a raise like below and having to patch the package manually (which we have to do currently).

```
MSC version unmatch: _MSC_VER: 1924 is expected
```

File: test_package/embedded/CMakeLists.txt [hunk 6]
```diff
@@ -4,7 +4,7 @@ add_executable(CreateEmbeddedSource
   CreateEmbeddedSource.cpp
 )
 
-target_link_libraries(CreateEmbeddedSource CONAN_PKG::ZLIB)
+target_link_libraries(CreateEmbeddedSource CONAN_PKG::zlib)
```
Comment (jmarrec) [inline test_package/embedded/CMakeLists.txt:9]: Update the test_package with the new lowercase convention in conan-center-index.

**Comment by jmarrec** [comment]
cf https://github.com/conan-io/conan-center-index/pull/2180

For now make sure you have the NREL remote before the conan-center-index one cf https://github.com/NREL/conan-openstudio-ruby/pull/28/commits/8e6af42339502f023110491e1bd6285ee00c0787

File: conanfile.py [hunk 7]
```diff
@@ -83,23 +83,32 @@ def requirements(self):
...
 
         if self.options.with_gdbm:
-            self.requires("gdbm/1.18.1@bincrafters/stable")
+            # NOTE: I have uploaded the gdbm/1.18.1 to the NREL remote
+            # with the status of this PR https://github.com/conan-io/conan-center-index/pull/2180
+            # at SHA https://github.com/conan-io/conan-center-index/pull/2180/commits/fad6b09ec294e8c0d186caea0c38bd6941dc0343
+            # So for now that'll only work if you have the NREL remote **before**
+            # the conan-center one...
+            # `conan remote update nrel https://api.bintray.com/conan/commercialbuilding/nrel --insert 0`
+            self.requires("gdbm/1.18.1")
```
Comment (jmarrec) [inline conanfile.py:97]: Important

**Comment by jmarrec** [comment]
Mac is having problems linking to gdbm... this needs to be resolved.

**Review by jmarrec — COMMENTED** [review]
> Mac is having problems linking to gdbm... this needs to be resolved.

It builds plenty fine on my own mac... great.

**Comment by jmarrec** [comment]
I uploaded my locally build gdbm/1.18.1 on mac apple-clang 10. Now travis for that version works. Jenkins for NREL/OpenStudio uses that osx version so at least that's good.

The only left failing is travis with xcode11.2, and it's due to gdbm too. I'm going to upgrade that to xcode11.4 to amtch the bincrafters templates and see.

Windows fails but just on x86 in the test_package since swig4.0.1 has trouble building on x86. The openstudio_ruby step works fine.

```
swig/4.0.1: autom4te.exe: need GNU m4 1.4 or later: /usr/bin/env m4
```

**Comment by jmarrec** [comment]
I'm trying to get Travis to spit the ext/gdbm/mkmf.log. https://travis-ci.org/github/NREL/conan-openstudio-ruby/jobs/711361148
So I'm canceling all other jobs, don't mind the red "X"s for now.

**Comment by jmarrec** [comment]
Comparing from a mac with apple-clang10 and the travis with apple-clang11:

```
"clang
-o
conftest
-I../../.ext/include/x86_64-darwin19
-I../.././include
-I../.././ext/gdbm
-I/Users/travis/.conan/data/gdbm/1.18.1/_/_/package/3c1fe12564616d853724cec6933b1304fd76a273/include
-D_XOPEN_SOURCE
-D_DARWIN_C_SOURCE
-D_DARWIN_UNLIMITED_SELECT
-D_REENTRANT


-O3
-ggdb3
-Wall
-Wextra
-Wno-unused-parameter
-Wno-parentheses
-Wno-long-long
-Wno-missing-field-initializers
-Wno-tautological-compare
-Wno-parentheses-equality
-Wno-constant-logical-operand
-Wno-self-assign
-Wunused-variable
-Wimplicit-int
-Wpointer-arith
-Wwrite-strings
-Wdeclaration-after-statement
-Wshorten-64-to-32
-Wimplicit-function-declaration
-Wdivision-by-zero
-Wdeprecated-declarations
-Wextra-tokens

-pipe
conftest.c

-L.
-L../..
-L/Users/travis/.conan/data/gdbm/1.18.1/_/_/package/3c1fe12564616d853724cec6933b1304fd76a273/lib
-L.
-fstack-protector
-L/usr/local/lib




-lruby.2.5.5-static
-framework
Foundation
-lgdbm

-lpthread
-lgmp
-ldl
-lobjc
"

Undefined symbols for architecture x86_64:
  "_libintl_dgettext", referenced from:
      __gdbm_get_bucket in libgdbm.a(bucket.o)
      __gdbm_write_bucket in libgdbm.a(bucket.o)
      __gdbm_split_bucket in libgdbm.a(bucket.o)
      _gdbm_strerror in libgdbm.a(gdbmerrno.o)
      _gdbm_db_strerror in libgdbm.a(gdbmerrno.o)
      __gdbm_alloc in libgdbm.a(falloc.o)
      _push_avail_block in libgdbm.a(falloc.o)
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```


Side by side diff:


image

**Comment by jmarrec** [comment]
Am reading "I get a linker error “undefined reference to libintl_gettext”: https://www.gnu.org/software/gettext/FAQ.html#integrating_undefined

**Comment by jmarrec** [comment]
Waaaaw tracking it down is fun...

Once the PR https://github.com/conan-io/conan-center-index/pull/2180 is merged and I can retrieve the conan-center-index package I **think** there is a chance I will no longer get the problem...

the conan-openstudio-ruby for travis xcode11.4 is building GDBM as there aren't any packages for it in the remote. And there is a difference in the configuration versus building ONLY gdbm:

Building only GDBM (log: https://c3i.jfrog.io/c3i/misc/logs/pr/2180/5/gdbm/1.18.1/16cd4d7fadaeba696a90befbf0a9ca41217695f0/create_stdout.txt (matching [profile.txt](https://c3i.jfrog.io/c3i/misc/logs/pr/2180/5/gdbm/1.18.1/16cd4d7fadaeba696a90befbf0a9ca41217695f0/profile.txt))

```
checking for GNU gettext in libc... no
checking for iconv... yes
checking for working iconv... yes
checking how to link with libiconv... -liconv
checking for GNU gettext in libintl... no
checking whether to use NLS... no
```

Our travis instance:

```
checking for GNU gettext in libc... no
checking for iconv... yes
checking for working iconv... yes
checking how to link with libiconv... /Users/travis/.conan/data/libiconv/1.16/_/_/package/f74366f76f700cc6e991285892ad7a23c30e6d47/lib/libiconv.a
checking for GNU gettext in libintl... yes
checking whether to use NLS... yes
checking where the gettext function comes from... external libintl
checking how to link with libintl... -lintl -Wl,-framework -Wl,CoreFoundation
```


**I did fix the xcode10 build by manually building GDBM and uploading the package on my old macbook pro with has xcode10, so I think once there is a package the same will happen**

**Comment by jmarrec** [comment]
Trying a new thing: I patched the gdbm recipe to pass a `--disable-nls` flag...

**Edit**: That seems to work.

**Comment by jmarrec** [comment]
Geez, Travis is failing at upload stage on mac now... 

```
ERROR: 405 Client Error: Not Allowed for url: https://api.bintray.com/conan/commercialbuilding/nrel/v1/files/nrel/openstudio_ruby/2.5.5/testing/0/package/e4da4946ea40704f5fd453ac71fc52ebc64f4605/0/conan_package.tgz
```
Appveyor looks like it's going to hit the 1hr limit...

But otherwise it seems to work....

**Comment by jmarrec** [comment]
I added libyaml to conan-center-index, now I need to see if I can use it. Back to testing....

https://github.com/conan-io/conan-center-index/pull/2245

**Comment by jmarrec** [comment]
Some of the windows appveyor instance appear to stop communicating at one point...

**Comment by jmarrec** [comment]
Interesting, I'm building conan-openstudio-ruby on debug on Windows (MSVC 2019), and apparently it's hitting plenty of bison exceptions, including where the appveyor debug builds appear to stop

bison_exception


If I keep clicking "Ignore" it'll keep going

I think the ruby configuration is testing whether things are supported or not, but bison being in Debug actually creates the popup windows and Appveyor doesn't click on them... The strange thing is why did it work before, though perhaps it IS a bison thing (I did upgrade bison as part of this PR)

**Comment by jmarrec** [comment]
I built manually the packages on Windows with the same config as CI, and uploaded that the NREL/testing

image


Will relaunch a build of https://github.com/NREL/OpenStudio/pull/4015 to make sure everything works then this one can merge.

**Comment by jmarrec** [comment]
My local windows machine when uploading package was marking the CI ones as outdated... LINE ENDINGS MATTERS for the recipe hash!! DAMN.



git config core.autocrlf input
git config core.eol lf
git checkout-index --force --all


https://docs.conan.io/en/latest/faq/using.html#packages-got-outdated-when-uploading-an-unchanged-recipe-from-a-different-machine

**Comment by jmarrec** [comment]
Configure git for a conan repo to use LF to avoid computing a different hash: https://docs.conan.io/en/latest/faq/using.html#packages-got-outdated-when-uploading-an-unchanged-recipe-from-a-different-machine

```
git config core.autocrlf input
git config core.eol lf
git checkout-index --force --all
git rm --cached -r .  # Remove every file from git's index.
git reset --hard      # Rewrite git's index to pick up all the new line endings.
```

Gitbash: check that it worked:

OK:

```
$ file x64-mingw32.patch
x64-mingw32.patch: unified diff output, ASCII text
```

NOT OK:

```
$ file x64-mingw32.patch
x64-mingw32.patch: unified diff output, ASCII text, with CRLF line terminators
```

**Comment by tijcolem** [comment]
@jmarrec Nice work getting the patch in place for win32 builds. Since you can build these locally I'll just ignore the appyevor build errors since they are hitting the timeout of 1 hour.   Is there anything else to attend to or can this be merged in? Once in, I'll merge in the https://github.com/NREL/OpenStudio/pull/4015.

**Comment by jmarrec** [comment]
@tijcolem I was waiting for a few CCI upstream Prs to go in related to m4 and bison. It seems some have dropped in an hour ago, I'll check tomorrow is everything is ok. THese should allow APPVEYOR to build debug again. My problem with local builds is that I couldn't match my msvc recipe shas to the unix travis ones. Peehaps this will help

**Comment by jmarrec** [comment]
With the new bison recipe instead of having Release working and Debug hanging, now we have none of them working...

cf https://github.com/conan-io/conan-center-index/pull/2334#issuecomment-669851045

**Comment by jmarrec** [comment]
Ok, I patched the upstream bison to remove some env variable. It should work now with 3.5.3 only, 3.7.1 is stricter and will throw with ruby's shitty configure stuff

**Comment by jmarrec** [comment]
https://github.com/conan-io/conan-center-index/pull/2438 was opened to address the bison thing. For some reason appveyor isn't picking up the bison packages I built manually on my windows machine (it should...) on some of the runs and does not manage to build bison itself... So we'll have to wait for CCI to produce packages for every configuration..

**Pull request merged**